include/fs/littlefs.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2019 Bolt Innovation Management, LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_FS_LITTLEFS_H_
+#define ZEPHYR_INCLUDE_FS_LITTLEFS_H_
+
+#include <zephyr/types.h>
+#include <kernel.h>
+#include <storage/flash_map.h>
+
+#include <lfs.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @brief Filesystem info structure for LittleFS mount */
+struct fs_littlefs {
+	/* Defaulted in driver, customizable before mount. */
+	struct lfs_config cfg;
```

Comment:
Would it be possible to change this to:
```c
const struct lfs_config *cfg;
```

and

1) extend configuration macros so it will be possible to setup `lfs_config` as `static const` structure at compile time and has it whole at a flash
2) add function

```c
int littlefs_config(struct fs_mount_t *mountp, struct lfs_config *cfg);
```
which will be runtime version (i.e. need to be called before `littlefs_mount`) ?

Reply:
For (1), no.  `lfs_config.{read,prog,sync}` are assigned within the driver to the functions the driver implements for a specific back-end.  The structure must be in RAM.

For (2), I don't understand who would be calling that function.  It's not in the FS API.

Reply:
> For (1), no. `lfs_config.{read,prog,sync}` are assigned within the driver to the functions the driver implements for a specific back-end. The structure must be in RAM.
> 
I don't see any specific back-end other than defined in lfs_api_xxx functions (i.e. functions that wraps flash_area_xxx functions) and even if there would be so, it certainly does not need to be in RAM.

> For (2), I don't understand who would be calling that function. It's not in the FS API.

Every concrete filesystem in Zephyr need specific setup and initialization before mount (for example in fs shell, there are mount functions for fat, nffs and littlefs. Those initialization API does not need to be FS API, it is specific to FS (so better than API is to use term SPI in meaning of 'service provider interface').

From use cases point of view there are two:
1.) static FS configuration - everything is known before build, i.e. flash parameters, sizes, offsets, buffers and other littlefs settings - this is common for most cases and target devices.
2.) runtime FS configuration that otherwise need to call second function - for example posix host utility configured by arguments from command line, or dynamic FS setup for micropython etc.

I just don't like to see wasting RAM for constant structures.

Reply:
It is not a constant structure given the way the FS API works.  See https://github.com/zephyrproject-rtos/zephyr/pull/17749/files#r310552307.

The shell really does nothing more than put some generic glue code that's specific to the shell around a call to `fs_mount()` given a pointer to a statically-defined mount point.  NFFS has to get the device name at runtime; fatfs and littlefs do not.  Generally there is no FS-specific runtime configuration.

Reply:
> It is not a constant structure given the way the FS API works. See https://github.com/zephyrproject-rtos/zephyr/pull/17749/files#r310552307.
> 
> The shell really does nothing more than put some generic glue code that's specific to the shell around a call to `fs_mount()` given a pointer to a statically-defined mount point. NFFS has to get the device name at runtime; fatfs and littlefs do not. Generally there is no FS-specific runtime configuration.

For static configurations you wouldn't be forced to call littlefs specific config function. However there are other cases where use of such interface is legitimate: format, chkdsk, etc.) without it, it would not even be possible to create utility like described in #17719 (without recompilation everytime FS parameters is changed).

Reply:
So in those special cases you need runtime configuration, which you can do with this as it stands.

It seems that all this is about trying to move 72 bytes of required configuration data from RAM into flash.  As the submitter and maintainer of this module my answer is: no.  It doesn't increase functionality, and it dramatically increases complexity because of the visibility and accessibility of information that would have to be collected at compile time.

It's up to you whether you want to approve, comment, or reject this PR.  If you want to figure out how to make this work later, that's on you.

Reply:
> figure out how to make this work later

I'd definitely recommend any reviewer to keep this option in mind. Suggesting an improvement may be good, but an answer "it can't be all done at once" should be a well expected response.

Reply:
> So in those special cases you need runtime configuration, which you can do with this as it stands.
> 
> It seems that all this is about trying to move 72 bytes of required configuration data from RAM into flash. As the submitter and maintainer of this module my answer is: no. It doesn't increase functionality, and it dramatically increases complexity because of the visibility and accessibility of information that would have to be collected at compile time.
> 
> It's up to you whether you want to approve, comment, or reject this PR. If you want to figure out how to make this work later, that's on you.

Yeah, its not my intention to stall this PR. As I said I agree with it as is.

Reply:
OK.  When the review is comment-only it's hard to tell where change requests fall on the spectrum from "it would be nice if..." to "if you don't change this I'll block it".

I usually try pretty hard to accommodate requests (I spent an hour trying to write up why I couldn't make partitions independently configurable before I figured out how to make it work), but this one just isn't going to work at this time.

@pkral78 if github lets you select "Approve" in the review box, or change the review type from "Comment" to "Approve", please do.  Comment-only approvals are nice, but the folks who merge won't see them.  If it won't let you do that, ok.  Thanks for your input; it's improved the solution significantly.